src/activity_handlers.cpp

```diff
@@ -2728,6 +2730,50 @@ void activity_handlers::try_sleep_finish( player_activity *act, player *p )
     act->set_to_null();
 }
 
+void activity_handlers::build_do_turn( player_activity *act, player *p )
+{
+    item *con_item = act->targets.front().get_item();
+    // this shouldn't happen, unfinished crafts should not dissappear during activity
+    if( !con_item ) {
+        add_msg( m_bad, "The marker item is no longer there, cancelling construction," );
+        p->cancel_activity();
+        return;
+    }
+    std::vector<construction> list_constructions = get_constructions();
+    const construction &built = list_constructions[act->index];
+
+    // item_counter represents the percent progress relative to the base batch time
+    // stored precise to 5 decimal places ( e.g. 67.32 percent would be stored as 6732000 )
+    const int old_counter = con_item->get_var( "construction_progress", 0 );
+
+    // Base moves for construction with no speed modifier or assistants
+    // Must ensure >= 1 so we don't divide by 0;
+    const double base_total_moves = std::max( 1, built.time );
+    // Current expected total moves, includes construction speed modifiers and assistants
+    const double cur_total_moves = std::max( 1, built.adjusted_time() );
+    // Delta progress in moves adjusted for current crafting speed
+    const double delta_progress = p->get_moves() * base_total_moves / cur_total_moves;
+    // Current progress in moves
+    const double current_progress = old_counter * base_total_moves / 10000000.0 +
+                                    delta_progress;
+    // Current progress as a percent of base_total_moves to 2 decimal places
+    con_item->set_var( "construction_progress",
+                       round( current_progress / base_total_moves * 10000000.0 ) );
+    p->set_moves( 0 );
+
+    con_item->set_var( "construction_progress",
+                       std::min( static_cast<int>( con_item->get_var( "construction_progress", 0 ) ), 10000000 ) );
+    std::string con_desc = string_format( _( "Unfinished task: %s. It is %d percent complete" ),
```

Comment:
Not really a reason to create this string every turn, how about just doing it when it needs to be displayed?

Reply:
Surely that would involve more code in more places to get all the times when it may be displayed? rather than just ensuring it's name is updated when it needs to be updated?

Reply:
Well you shouldn't ever be able to actually see this item right?  So it should just require adding code to the iexamine you have asking to continue.  Actually you probably want to make `trap::name()` check if the trap is an unfinished construction and if so look for the pseudo item with progress data and display that.

Reply:
You can see it in two scenarios, with advanced inventory, or when you stand over it and press ,
You cant move it or pick it up, like its a liquid, but it displays the name.
That feels better than just not displaying it at all, idk.